### PR TITLE
Handle %g in printf (fix #11599)

### LINKED ISSUE #11599: Implement printf %g format specifier
The printf `%g` format specifier is [currently unimplemented](https://github.com/JuliaLang/julia/blob/3ff49ff39bd56a4928ae4341e60030c6b170e7a1/base/printf.jl#L647-L649).  It would be great to be able to use it in Julia code.

Is there a specification somewhere for what this *should* do?  The best I can find is [python's documentation on the subject](https://docs.python.org/3.4/library/string.html#format-specification-mini-language), which contains documentation on how `g` works there.

### PR DESCRIPTION
**Pull request by dfannius**
Fixes issue  #11599.

This was more involved than expected, even though the spec of `%g` is mostly just to delegate to `%e` or `%f`, because the `@printf` implementation makes some fundamental assumptions that break in the case of `%g`, including the assumptions that:
- we know the precision at macro-expansion time (not true for `%g` because it depends on the exponent)
- we can calculate the correct left padding before actually printing the value (not true for `%g` because trailing `0`s are removed)

I prioritized changing the existing code as little as possible; otherwise I might have refactored a little more. I did add optional arguments to `print_fixed()` and `gen_e()`. One could imagine an implementation of `@printf` that put off more explicit work until runtime in the expectation that it would be optimized away in most cases, in order to enable more code-sharing between `%f` and `%g`, but if nobody has ever even tried to use `%g` until now, that seems like overkill.

This is my first pull request for core Julia, so criticisms are especially welcome.

### CODE REVIEW

**Comment by ScottPJones** [comment]
One thing I'd say is to have a lot more test cases (not just similar to the %e and %f cases). It would be good if all of your new code is completely covered when running the unit tests
(a number of people are working on making the code coverage more complete).
Nice to see this!

**Comment by dfannius** [comment]
Good call. I was going to add more tests but felt a little silly given how minimal the tests for `%e` and `%f` were. I added a few more tests that cover all the branches of the obvious conditionals, which turned up one outright bug and one case where I hadn't read the spec closely enough. The pull request has been updated.

**Comment by dfannius** [comment]
I should also mention, with respect to issue #11942, that this code handles `NaN` and `Inf` for `%g` the same way that Julia currently handles them for `%e` and `%f` (in fact, it calls the same code).

File: base/printf.jl [hunk 1]
```diff
@@ -165,7 +165,25 @@ function pad(m::Int, n, c::Char)
...
+                $i -= 1
+            end
+        end
+    end
+end
+
+function print_fixed(out, precision, pt, ndigits, trailingzeros=true)
+    # @show( precision )
+    # @show( pt )
+    # @show( ndigits )
```
Comment (tkelman) [inline base/printf.jl]: should delete this debugging code

Comment (dfannius) [inline base/printf.jl]: Whoops, thanks for the catch. Removed.

**Comment by garrison** [comment]
This is exciting.  Would you be willing to rebase?  `test/strings.jl` is gone as of 38c6925, so the tests should go in one of the new files.

**Comment by dfannius** [comment]
Glad to see that someone's excited! I'll try to rebase this weekend.

**Comment by StefanKarpinski** [comment]
I'm excited about this too. Code review is daunting, however. If you rebase I do promise to review it :-)

**Comment by StefanKarpinski** [comment]
I spent about ten minutes on trying to get this build to go again, but the AppVeyor interface has defeated me. It refuses to restart the build, telling me to contact support. The support contact page requires you authenticate separately from the main app, but only supports password and Google auth, while my account is setup to authenticate with GitHub – which isn't an option. I also can't figure out where in the interface one goes to give others admin rights to the repo or how to create an email / password login (changing the password requires entering an old password, which I don't have).

**Comment by dfannius** [comment]
Rebased successfully (I think!) and moved the tests into `printf.jl`. The reference output for all the tests came from clang.

**Comment by jakebolewski** [comment]
bump

**Comment by StefanKarpinski** [comment]
I've built and tested this branch a bit and it seems good but I want to check the performance a bit more.

**Comment by StefanKarpinski** [comment]
Good lord, the code gen is crazy:

```jl
julia> macroexpand(:(@sprintf("%.6g", 1234567)))
:(let
        #65#out = Base.Printf.IOBuffer()
        #66###x#6256 = 1234567
        local #59#neg, #60#pt, #61#len, #62#exp, #63#do_out, #64#args
        if Base.Printf.isfinite(#66###x#6256)
            (#63#do_out,#64#args) = Base.Printf.ini_dec(#65#out,#66###x#6256,7,"",0,6,'g')
            if #63#do_out
                (#61#len,#60#pt,#59#neg) = #64#args
                #62#exp = Base.Printf.-(#60#pt,1)
                #67#do_f = (6 Base.Printf.> #62#exp Base.Printf.>= -4)
                if #67#do_f
                    #68#fprec = Base.Printf.-(6,Base.Printf.+(#62#exp,1))
                    (#63#do_out,#64#args) = Base.Printf.fix_dec(#65#out,#66###x#6256,"",0,#68#fprec,Base.Printf.-('g',1))
                    if #63#do_out
                        (#61#len,#60#pt,#59#neg) = #64#args
                        #69#padding = Base.Printf.nothing
                        #70#width = 0
                        #71#tmpout = Base.Printf.IOBuffer()
                        if (#68#fprec Base.Printf.> 0) # line 713:
                            Base.Printf.print_fixed(#71#tmpout,#68#fprec,#60#pt,#61#len,false)
                        else  # line 715:
                            Base.Printf.write(#71#tmpout,Base.Printf.pointer(Base.Printf.DIGITS),#61#len) # line 716:
                            while (#60#pt Base.Printf.>= #61#len += 1) # line 716:
                                Base.Printf.write(#71#tmpout,'0')
                            end
                        end
                        #72#tmpstr = Base.Printf.takebuf_string(#71#tmpout)
                        if (#68#fprec Base.Printf.> 0) # line 719:
                            #70#width -= Base.Printf.length(#72#tmpstr)
                        end
                        if #59#neg # line 723:
                            #70#width -= 1
                        end
                        if (#70#width Base.Printf.>= 1) # line 725:
                            #69#padding = #70#width
                        end
                        if (#69#padding Base.Printf.!= Base.Printf.nothing) # line 730:
                            begin  # printf.jl, line 171:
                                if (#70#width Base.Printf.<= 1) # line 172:
                                    (#69#padding Base.Printf.> 0) && Base.Printf.write(#65#out,' ')
                                else  # line 174:
                                    #73###i#6257 = #69#padding # line 175:
                                    while (#73###i#6257 Base.Printf.> 0) # line 176:
                                        Base.Printf.write(#65#out,' ') # line 177:
                                        #73###i#6257 -= 1
                                    end
                                end
                            end
                        end
                        #59#neg && Base.Printf.write(#65#out,'-')
                        Base.Printf.write(#65#out,#72#tmpstr)
                    end
                else
                    #74###x#6258 = #66###x#6256
                    begin
                        (#63#do_out,#64#args) = Base.Printf.ini_dec(#65#out,#74###x#6258,6,"",0,5,'g')
                        if #63#do_out
                            (#61#len,#60#pt,#59#neg) = #64#args
                            #62#exp = Base.Printf.-(#60#pt,1)
                            #59#neg && Base.Printf.write(#65#out,'-')
                            Base.Printf.write(#65#out,Base.Printf.DIGITS[1])
                            begin
                                #75#endidx = 6
                                begin
                                    while (#75#endidx Base.Printf.> 1) && (Base.Printf.DIGITS[#75#endidx] Base.Printf.== '0') # line 470:
                                        #75#endidx -= 1
                                    end
                                    if (#75#endidx Base.Printf.> 1) # line 473:
                                        Base.Printf.write(#65#out,'.') # line 474:
                                        Base.Printf.write(#65#out,Base.Printf.+(Base.Printf.pointer(Base.Printf.DIGITS),1),Base.Printf.-(#75#endidx,1))
                                    end
                                end
                            end
                            Base.Printf.write(#65#out,'e')
                            Base.Printf.print_exp_e(#65#out,#62#exp)
                        end
                    end
                end
            end
        else
            Base.Printf.write(#65#out,begin  # printf.jl, line 145:
                    if Base.Printf.isnan(#66###x#6256)
                        "NaN"
                    else
                        if (#66###x#6256 Base.Printf.< 0)
                            "-Inf"
                        else
                            "Inf"
                        end
                    end
                end)
        end
        Base.Printf.nothing
        Base.Printf.takebuf_string(#65#out)
    end)
```

But that's really just a logical extension of how this `@printf` stuff works in general, which is not this PR's fault. So I think we're good with this.

**Comment by jakebolewski** [comment]
why is the trailing `Base.Printf.nothing` there?

**Comment by dfannius** [comment]
@StefanKarpinski Yeah, it's twice as bad as the usual generated code for `printf` because it can't know until runtime whether to use the `%f`- or `%e`-like format, so the code for both has to stick around. The one performance impact on other code is that the `%e` code (`gen_e`) became slightly more general and acquired an extra parameter that it has to check a couple of times (to see whether it is being used directly by `%e` or indirectly by `%g`), but that overhead seemed pretty trivial.

**Comment by StefanKarpinski** [comment]
Totally makes sense – this is why I didn't implement `%g` in the first place ;-)

Isn't the overhead in `%e` entirely compile time?

**Comment by simonster** [comment]
Is there really a performance advantage to doing all of this codegen rather than calling a function?

**Comment by StefanKarpinski** [comment]
The `Base.Printf.nothing` comes from [here](https://github.com/JuliaLang/julia/blob/66f336d37e123482c9fe5201750564537c420cbe/base/printf.jl#L32) – totally unrelated and pre-existing. It was originally to make sure that `@printf` returned nothing; I'm not sure if it's necessary anymore, but it's tangential in any case.

**Comment by dfannius** [comment]
@StefanKarpinski I thought it was happening at macro expansion time, which is still happening on the user's clock, but in any case it's pretty trivial. I guess that means that it's happening once per time that `%e` is encountered in the code, not once per time that `%e` is executed, which can be a big difference.

**Comment by dfannius** [comment]
@simonster Stefan talked about this a bit here: http://stackoverflow.com/questions/19783030/in-julia-why-is-printf-a-macro-instead-of-a-function

**Comment by StefanKarpinski** [comment]
> Is there really a performance advantage to doing all of this codegen rather than calling a function?

Open question. This `@printf` implementation is a total monster and needs to be redone pretty badly.

Interesting historical tidbit... this `@printf` implementation was inspired by [this PyPy blog post](http://morepypy.blogspot.com/2011/08/pypy-is-faster-than-c-again-string.html), which claimed that by specializing on the constant format string, they'd gotten `sprintf("%d %d", i, i)` twice as fast as C. Sounds great, right? So I set out to write a macro-based printf that specialized the hell out of its code based on the constant format string. Result: no faster than C. Hmm...

After some investigation I found that the blog post is completely wrong about why PyPy is faster than C: they're not getting performance from specializing on the format string – they're getting it from specializing on the fact that they're printing the same number twice and can decode it only once. Since decoding the number to decimal is where almost all of the time is spent (cf writing bytes to memory), PyPy is twice as fast as C. Of course, a clever C compiler could also notice this and specialize for it. But how often do you print the same exact number twice?

Also, while PyPy was faster than C (and Julia) at `sprintf`, if you changed the benchmark to `printf` – i.e. actually printing to a stream instead of just writing to a string – then it was 7x slower than C.

Ancient discussion thread [here](https://groups.google.com/forum/#!topic/julia-dev/L0iMxbYBMR8/discussion). I think that Jeff, Viral and I were the only people on the list at the time.

**Comment by StefanKarpinski** [comment]
I'm going to merge and we can discuss how to make our `printf` implementation better elsewhere.

**Comment by StefanKarpinski** [comment]
Thanks for the excellent (and somewhat crazy) `%g` implementation, @dfannius!

**Comment by garrison** [comment]
Awesome, excited to see this in 0.4!

**Pull request merged**